runtime/shared_common/ClasspathManagerImpl2.cpp

```diff
@@ -1231,8 +1231,15 @@ SH_ClasspathManagerImpl2::storeNew(J9VMThread* currentThread, const ShcItem* ite
 		U_16 pathLen = 0;
 		const char* path = cpi->itemAt(i)->getLocation(&pathLen);
 		U_8 isToken = (cpi->getType()==CP_TYPE_TOKEN);
+		U_8 flags = 0;
 
-		if (!cpeTableUpdate(currentThread, path, pathLen, i, itemInCache, isToken, isLastItem, cachelet)) {
+		/* force a timestamp check once for URLClasspath */
+		if (J9_ARE_NO_BITS_SET(*_runtimeFlagsPtr, J9SHR_RUNTIMEFLAG_ENABLE_CLASSPATH_TIMESTAMP_CHECK)
+		&& (CP_TYPE_CLASSPATH == cpi->getType())) {
+			flags = (CPM_ZIP_OPEN | CPM_ZIP_FORCE_CHECK_TIMESTAMP);
```

Comment:
The flag is set to CPM_ZIP_OPEN | CPM_ZIP_FORCE_CHECK_TIMESTAMP when we are storing a new class path into the shared cache. What about the case that the class path already exists in the shared cache ?

Reply:
J9SHR_RUNTIMEFLAG_ENABLE_CLASSPATH_TIMESTAMP_CHECK has been checked when calling notifyClasspathChange3 (open or close state)
```
Java_com_ibm_oti_shared_SharedClassURLClasspathHelperImpl_notifyClasspathChange3(JNIEnv* env, jobject thisObj, jint helperID, jobject classLoaderObj, jobjectArray urlArrayObj, jint urlCount, jboolean isOpen)
{
...
	if (vm->sharedClassConfig->runtimeFlags & J9SHR_RUNTIMEFLAG_ENABLE_CLASSPATH_TIMESTAMP_CHECK) {
		Trc_JCL_com_ibm_oti_shared_SharedClassURLClasspathHelperImpl_notifyClasspathChange3_ExitEnableCheck(env);
		return;
	}
...
	ALWAYS_TRIGGER_J9HOOK_VM_ZIP_OPEN_STATE(vm->hookInterface, vm, classloader, cpEntries, entryCount, isOpen);

```

Reply:
I guess the change here does nothing. As CPM_ZIP_OPEN | CPM_ZIP_FORCE_CHECK_TIMESTAMP should already be set earlier when initializing the class loader where SH_ClasspathManagerImpl2::notifyClasspathEntryStateChange() is called that adds the jar to the hashtable and set CPM_ZIP_OPEN | CPM_ZIP_FORCE_CHECK_TIMESTAMP.

```
SH_ClasspathManagerImpl2::notifyClasspathEntryStateChange() {
...
          header = cpeTableLookup(currentThread, pathChar, pathLen, 0);			/* 0 = isToken. Certainly won't be a token. */
	if (header == NULL) {
		header = cpeTableAddHeader(currentThread, pathChar, pathLen, NULL, 0);
	}
	
	if (header) {
		UDATA oldFlags = header->_flags;
		
		switch(newState) {
		case J9ZIP_STATE_OPEN :
			/* CMVC 201090: Note that this step clears out any existing flags on the container. */
			header->_flags = (CPM_ZIP_OPEN | CPM_ZIP_FORCE_CHECK_TIMESTAMP);
			break;
```

Reply:
I think the question above is whether the changes (set up the flags) in storeNew() is redundant as we always set up the flags when creating a new URLClassloader (via notifyClasspathEntryStateChange). I am wondering whether there is any other scenario in which case the only way to set them up is via storeNew().

Reply:
Ah, this piece of code will be triggered either the JVM is storing or has found a new classpath entry in the shared cache. 

It is possible that the class loader can update its class path after initialization.